File: src/main/java/com/kit/log/ZogFile.java
/*
 * Copyright (c) 2018.
 * Author：Zhao
 * Email：joeyzhao1005@gmail.com
 */

package com.kit.log;

import android.content.Context;
import android.util.Log;

import androidx.annotation.StringDef;

import com.kit.app.application.AppMaster;
import com.kit.config.AppConfig;
import com.kit.utils.AppUtils;
import com.kit.utils.DateUtils;
import com.kit.utils.FileUtils;
import com.kit.utils.RxUtils;
import com.kit.utils.StringUtils;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author joeyzhao
 */
public final class ZogFile {
    /**
     * log默认存储目录
     */
    private static String defaultDir;

    /**
     * log存储目录
     */
    private static String dir;


    private static List<String> tagFilter;


    public static void setDir(final String dir) {
        if (StringUtils.isEmptyOrNullStr(dir)) {
            ZogFile.dir = null;
        } else {
            ZogFile.dir = dir.endsWith(File.separator) ? dir : dir + File.separator;
        }
    }

    public static void setDir(final File dir) {
        ZogFile.dir = dir == null ? null : dir.getAbsolutePath() + File.separator;
    }

    public static void setTagFilter(String... tagFilter) {
        ZogFile.tagFilter = Arrays.asList(tagFilter);
    }

    public static String getDir() {
        if (StringUtils.isEmptyOrNullStr(dir)) {
            defaultDir = AppMaster.getInstance().getAppContext().getExternalCacheDir() + File.separator + "logs";
            dir = defaultDir;
        }

        return dir;
    }

    public static String getLogFilePath() {
        return getDir() + File.separator + DateUtils.getCurrDateFormat("yyyyMMddHHmmss") + ".log";
    }

    public static void add(@Priority String priority, String tag, String logStr) {
        if (tagFilter == null || tagFilter.contains(tag)) {
            add(priority, "【" + tag + "】 " + logStr);
        }
    }

    public static void add(String logStr) {
        Log.e(INFO, logStr);
    }

    public static void add(@Priority String priority, String logStr) {
        AppConfig.IAppConfig appConfig = AppConfig.getAppConfig();
        if (appConfig == null || !appConfig.isShowLog()) {
            return;
        }

        if (!FileUtils.isExists(getDir())) {
            if (!FileUtils.mkDir(getDir())) {
                return;
            }
        }

        String filePath = getLogFilePath();
        File file = new File(filePath);
        boolean isCreated = true;
        if (!file.exists()) {
            try {
                isCreated = file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        if (!isCreated) {
            return;
        }

        RxUtils.io(null, new RxUtils.RxSimpleTask() {
            @Override
            public Object doSth(Object... objects) {
                try {
                    FileWriter writer = new FileWriter(filePath, true);
                    BufferedWriter bw = new BufferedWriter(writer);
                    String log = " 【" + DateUtils.getCurrDateFormat("HH:mm:ss") + "】 " + priority + " " + logStr + "\r\n";
                    bw.write(log);
                    bw.close();
                    writer.close();
                } catch (IOException e) {
                }

                return getDefault();
            }

            @Override
            public Object getDefault() {
                return "";
            }
        });
    }

    @StringDef({ERROR, INFO})
    @Retention(RetentionPolicy.SOURCE)
    @interface Priority {
    }

    public static final String ERROR = "ERROR";
    public static final String INFO = "INFO";

}

File: src/main/java/com/kit/log/ZogFile.kt
/*
 * Copyright (c) 2018.
 * Author：Zhao
 * Email：joeyzhao1005@gmail.com
 */
package com.kit.log

import android.util.Log
import androidx.annotation.StringDef
import com.kit.app.application.AppMaster
import com.kit.config.AppConfig
import com.kit.utils.DateUtils
import com.kit.utils.FileUtils
import com.kit.utils.StringUtils
import com.kit.utils.log.Zog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy
import java.util.*

/**
 * @author joeyzhao
 */
object ZogFile {
    /**
     * log默认存储目录
     */
    private var defaultDir: String? = null
    /**
     * log存储目录
     */
    private var dir: String? = null
    private var tagFilter: List<String>? = null
    fun setDir(dir: String) {
        if (StringUtils.isEmptyOrNullStr(dir)) {
            ZogFile.dir = null
        } else {
            ZogFile.dir = if (dir.endsWith(File.separator)) dir else dir + File.separator
        }
    }

    fun setDir(dir: File?) {
        ZogFile.dir = if (dir == null) null else dir.absolutePath + File.separator
    }

    fun setTagFilter(vararg tagFilter: String?) {
        ZogFile.tagFilter = Arrays.asList<String>(*tagFilter)
    }

    fun getDir(): String? {
        if (StringUtils.isEmptyOrNullStr(dir)) {
            defaultDir = AppMaster.getInstance().appContext.externalCacheDir.toString() + File.separator + "logs"
            dir = defaultDir
        }
        return dir
    }

    @JvmStatic
    val logFilePath: String
        get() = getDir() + File.separator + DateUtils.getCurrDateFormat("yyyyMMdd") + ".log"

    fun add(@Priority priority: String?, tag: String, logStr: String) {
        if (tagFilter == null || tagFilter!!.contains(tag)) {
            add(priority, "【$tag】 $logStr")
        }
    }

    fun add(logStr: String) {
        Log.e(INFO, logStr)
    }

    fun add(@Priority priority: String?, logStr: String?) {
        GlobalScope.launch(Dispatchers.IO) {
            val appConfig = AppConfig.getAppConfig()
            if (appConfig == null || !appConfig.isShowLog) {
                return@launch
            }

            if (!FileUtils.isExists(getDir())) {
                if (!FileUtils.mkDir(getDir())) {
                    return@launch
                }
            }

            val filePath = logFilePath
            val file = File(filePath)
            var isCreated = true
            if (!file.exists()) {
                try {
                    isCreated = file.createNewFile()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

            if (!isCreated && !file.exists()) {
                return@launch
            }

            try {
                val writer = FileWriter(filePath, true)
                val bw = BufferedWriter(writer)
                val log = " 【" + DateUtils.getCurrDateFormat("HH:mm:ss") + "】 " + priority + " " + logStr + "\r\n"
                bw.write(log)
                bw.close()
                writer.close()
            } catch (e: Exception) {
                Zog.showException(e)
            }

        }
    }

    const val ERROR = "ERROR"
    const val INFO = "INFO"

    @StringDef(ERROR, INFO)
    @Retention(RetentionPolicy.SOURCE)
    internal annotation class Priority
}